archiver/archive.go

```diff
@@ -0,0 +1,143 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package archiver
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/joeshaw/multierror"
+
+	"github.com/pkg/errors"
+)
+
+// PackageProperties defines properties describing the package. The structure is used for archiving.
+type PackageProperties struct {
+	Name    string
+	Version string
+	Path    string
+}
+
+// ArchivePackage method builds and streams an archive with package content.
+func ArchivePackage(w io.Writer, properties PackageProperties) (err error) {
+	gzipWriter := gzip.NewWriter(w)
+	tarWriter := tar.NewWriter(gzipWriter)
+	defer func() {
+		var me multierror.Errors
+
+		if err != nil {
+			me = append(me, err)
+		}
+
+		err = tarWriter.Close()
+		if err != nil {
+			me = append(me, errors.Wrapf(err, "closing tar writer failed"))
+		}
+
+		err = gzipWriter.Close()
+		if err != nil {
+			me = append(me, errors.Wrapf(err, "closing gzip writer failed"))
+		}
+
+		if me != nil {
+			err = me.Err()
+		}
+	}()
+
+	rootDir := fmt.Sprintf("%s-%s", properties.Name, properties.Version)
+
+	err = filepath.Walk(properties.Path, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		relativePath, err := filepath.Rel(properties.Path, path)
+		if err != nil {
+			return errors.Wrapf(err, "finding relative path failed (packagePath: %s, path: %s)", properties.Path, path)
+		}
+
+		if relativePath == "." {
+			return nil
+		}
+
+		header, err := buildArchiveHeader(info, filepath.Join(rootDir, relativePath))
+		if err != nil {
+			return errors.Wrapf(err, "building archive header failed (path: %s)", relativePath)
+		}
+
+		err = tarWriter.WriteHeader(header)
+		if err != nil {
+			return errors.Wrapf(err, "writing header failed (path: %s)", relativePath)
+		}
+
+		if !info.IsDir() {
+			err = writeFileContentToArchive(path, tarWriter)
+			if err != nil {
+				return errors.Wrapf(err, "archiving file content failed (path: %s)", path)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "processing package path '%s' failed", properties.Path)
+	}
+
+	err = tarWriter.Flush()
+	if err != nil {
+		return errors.Wrap(err, "flushing tar writer failed")
+	}
+
+	err = gzipWriter.Flush()
+	if err != nil {
+		return errors.Wrap(err, "flushing gzip writer failed")
+	}
+	return nil
+}
+
+func buildArchiveHeader(info os.FileInfo, relativePath string) (*tar.Header, error) {
+	header, err := tar.FileInfoHeader(info, "")
+	if err != nil {
+		return nil, errors.Wrapf(err, "reading file info header failed (info: %s)", info.Name())
+	}
+
+	header.Name = relativePath
+	if info.IsDir() {
+		header.Name = header.Name + "/"
```

Comment:
Is this `/` dependent on the platfrom or always `/`?

Reply:
I've added an additional condition to check the `/`.